feat: Sapphire guide (#711)
discordjs/guide: guide/sapphire/first-command.md

# Your first command

Now that you have created the `commands` folder, you're ready to make your first command file! In Sapphire, we usually define one command per file, but you can also define groups of them. Hop over to your `commands` folder, and create a new file, we'll show how to make a ping command, so we will name this `ping.js`. This is going to be a simple command that replies to a message and then edits with the elapsed time. We'll go into arguments and other features later.

Once you have your file, it's time to get started!

## Creating your command class

Before you do anything, at the start of your file, you're going to need to require Sapphire Framework again, specifically its Command class.

```js
const { Command } = require('@sapphire/framework');
```

Commands are classes that are exported in any of the following ways:

> ```js

Reviewer: Maybe this can have only commonjs as most parts of guide follow common js and having two of them *might* confuse the user on which to use

// CommonJS
module.exports = class MainExportedFoo extends Piece {};
exports.default = class DefaultExportedFoo extends Piece {};
exports.myCommand = class CustomExportedFoo extends Piece {};

// ECMAScript Modules
export default class ExportedDefaultFoo extends Piece {}
export class ExportedFoo extends Piece {}

> ```

Reviewer: Maybe this should only have the CommonJS way as that is used mostly through the guide and the user might be confused on which to choose?

Author: @Danktuary @almostSouji Sapphire's loader supports both CJS and ESM (even mixed within the same project), as well as it supports multiple exports within the same file, which differs from Commando, so I don't really have a good reference at how should we display the possibilities.

Furthermore, all of Sapphire (as well as discord.js) support ESM exports, so the ESM integration here is full.

Any suggestions regarding how to make this less confusing/easier to understand?

Reviewer 2: For now, use the built-in `CodeGroup` and `CodeGroupItem` components: https://vuepress2.netlify.app/reference/default-theme/markdown.html#custom-containers

They're not used anywhere else yet, so see how well it works with ESLint (e.g. test with a missing semicolon). If it doesn't lint, try adding newlines before and after the codeblocks.

Author: It lints correctly, and it also works beautifully, thanks!
![image](https://user-images.githubusercontent.com/24852502/126479538-47e166a9-da96-4c52-bcb1-c07317b9b188.png)

# Your first command

Now that you have created the `commands` folder, you're ready to make your first command file! In Sapphire, we usually define one command per file, but you can also define groups of them. Hop over to your `commands` folder, and create a new file, we'll show how to make a ping command, so we will name this `ping.js`. This is going to be a simple command that replies to a message and then edits with the elapsed time. We'll go into arguments and other features later.

Once you have your file, it's time to get started!

## Creating your command class

Before you do anything, at the start of your file, you're going to need to require Sapphire Framework again, specifically its Command class.

:::: code-group
::: code-group-item CommonJS
```js
const { Command } = require('@sapphire/framework');
```
:::
::: code-group-item ESM
```js

> import { Command } from '@sapphire/framework';

Author: I think ESLint complains due to duplicated identifiers in here, since exported members in ESM are accessible with their name, unlike in CommonJS.

> import { Command } from '@sapphire/framework';

Reviewer 2: I suppose the other exports could be commented out, leaving only the default export, or an `<!-- eslint-skip -->` could be used.

> import { Command } from '@sapphire/framework';

Author: Commenting out also works, in a way it also shows the other possibilities. 👍🏼

```
:::
::::

Commands are classes that are exported in any of the following ways:

:::: code-group
::: code-group-item CommonJS
```js
module.exports = class Foo extends Piece {};
exports.default = class Foo extends Piece {};
exports.myCommand = class Foo extends Piece {};
```
:::
::: code-group-item ESM
```js
export default class DefaultFoo extends Piece {}
export class Foo extends Piece {}
```
:::
::::

> Because its spread usage, we will be using `module.exports` in all the guides, so let's create the class and set `module.exports` to it!

Reviewer 2: ```suggestion
`module.exports` will be used for these pages. Let's begin populating the command class:
```

```js
module.exports = class PingCommand extends Command {
	constructor(context) {
		super(context, {
			aliases: ['pong'],
			description: 'Tests the latency.',
		});
	}
};
```

Don't let this scare you; it's straightforward.

- `context` is an object that contains some file metadata required by the `Command` and `Piece` (which `Command` extends) require in order to function.
- `name` is, by default, the name of the file without the extension, e.g. `ping.js` becomes `ping`, so we don't need to define it.
- `aliases` are other ways users can call the command. You can have as many as you want!
- `description` is the help text displayed when somebody uses a help command.

> There are many more properties you can use, but their sections will explain those.

Reviewer 3: Something like:
```suggestion
There are many more properties you can use. The upcoming sections will explain those.
```

> Don't let this scare you; it's straightforward.
> 
> - `context` is an object that contains some file metadata required by the `Command` and `Piece` (which `Command` extends) require in order to function.
> - `name` is, by default, the name of the file without the extension, e.g. `ping.js` becomes `ping`, so we don't need to define it.
> - `aliases` are other ways users can call the command. You can have as many as you want!
> - `description` is the help text displayed when somebody uses a help command.
> 
> There are many more properties you can use. The upcoming sections will explain those.

Reviewer 2: ```suggestion
An overview of what's defined in the constructor:

- `context`: an object that contains file metadata required by the `Command` and `Piece` classes (which `Command` extends) in order to function.
- `name`: by default, the name of the file without the extension, i.e. `ping.js` becomes `ping`, so there's no need to define it.
- `aliases`: other ways users can call the command. You can have as many as you want!
- `description`: the text displayed when the help command is used.

There are many other properties available, explained in upcoming sections.
```

Author: Perhaps "required by the `Command` and `Piece` classes (which `Command` extends)" should be a little rewritten here, as the note in parenthesis is meanat to be next to "Piece", unless that's valid.

Reviewer 3: So,
```suggestion
An overview of what's defined in the constructor:
`context`: an object containing file metadata that the `Command` and `Piece` (which is extended by `Command`) classes require to function.
- `name`: by default, the name of the file without the extension, i.e. `ping.js` becomes `ping`, so there's no need to define it.
- `aliases`: other ways users can call the command. You can have as many as you want!
- `description`: the text displayed when the help command is used.
There are many other properties available, explained in upcoming sections.
```
?

Author: Maybe "which is extended by the former"?

Reviewer 2: I updated it to "an object that contains file metadata required by the `Piece` class (which `Command` extends)". This seems more straightforward and less wordy

> ## Creating your run method
> 
> The next thing you're going to need is a `run` method. This should go right below the command's constructor. Inside, you'll run the command's logic:

Reviewer 2: ```suggestion
## Creating the `run` method

Commands have a `run` method, which executes the command logic. Define this below the command's constructor:
```

<!-- eslint-disable constructor-super -->

> ```js {6-8}

Reviewer 2: ```suggestion
```js {7-9}
```

module.exports = class PingCommand extends Command {
	constructor(context) {
		// ...
	}

	async run(message) {
		const response = await message.channel.send('Ping...');
		const latency = response.createdTimestamp - message.createdTimestamp;
		await response.edit(`Pong! Took me ${latency}ms.`);
	}
};
```

<!-- eslint-enable constructor-super -->

> As you can see, the `run` method is simply the code you want the bot to run when somebody uses the command. This code can be anything you can do in core discord.js, as Sapphire Framework is simply an extension.
> 
> <!-- TODO(kyranet): Add a mention for editable commands once I make the plugin for it. -->
> 
> Now fire up the bot as usual and use your command! It should automatically be `@bot ping` to use it.

Reviewer 2: ```suggestion
Any discord.js code can be executed here, since Sapphire Framework is an extension. The command can be triggered with `@bot ping` or `@bot pong`.

<!-- TODO(kyranet): Add a mention for editable commands once I make the plugin for it. -->
```

Author: Shouldn't this be "since Sapphire Framework is an extension **of it**"? Or is it redundant?

Reviewer 3: > Shouldn't this be "since Sapphire Framework is an extension **of it**"? Or is it redundant?
I think both are correct, but this is better?

Reviewer 2: You're right, updated my suggestion
